### DISCUSSION TITLE
Does this rule engine use Rete or any other algorithm for processing efficiency?

### DISCUSSION THREAD

**Discussion by ravimpatel** [microsoft/RulesEngine#192] [Q&A]
Many rules engines (e.g. Biztalk, NRules etc.) use Rete algorithm for efficiency. Interested to know how this rule engine handles scenarios where there are many rules to be applied on many objects.

**Comment by ghost** [comment]
I took a really quick look at the code and it looks like it runs the rules in order. Though the magic seems to be in the pre-compiling & caching of the ruleset's by workflow. So as you add workflows to the engine instance, it will pre-emptively compile the expression trees necessary to evaluate the various predicates.

https://github.com/microsoft/RulesEngine/blob/4a2b345fe9e52ac5c4a856057b7815fd4df990da/src/RulesEngine/RulesEngine.cs#L332

![image](https://user-images.githubusercontent.com/28452123/150410243-8b95bcef-febe-42e4-ae94-b640e94c1c88.png)
